parsec/core/mountpoint/manager.py

```diff
@@ -222,23 +234,74 @@ async def get_timestamped_mounted(self):
             if workspace_id_and_ts[1] is not None
         }
 
+    def is_workspace_mounted(self, workspace_id, timestamp=None):
+        return (workspace_id, timestamp) in self._mountpoint_tasks
+
 
 @asynccontextmanager
 async def mountpoint_manager_factory(
-    user_fs, event_bus, base_mountpoint_path, *, enabled=True, debug: bool = False, **config
+    user_fs,
+    event_bus,
+    base_mountpoint_path,
+    *,
+    debug: bool = False,
+    mount_all: bool = False,
+    mount_on_workspace_created: bool = False,
+    mount_on_workspace_shared: bool = False,
+    exclude_from_mount_all: list = (),
 ):
-    config["debug"] = debug
+    config = {"debug": debug}
 
     runner = get_mountpoint_runner()
 
+    def on_new_workspace(event, new_entry, previous_entry=None):
+        # Only a role update
+        if previous_entry is not None:
+            return
+
+        async def target():
+            try:
+                await mountpoint_manager.mount_workspace(new_entry.id)
+            except MountpointAlreadyMounted:
+                pass
+            except Exception:
+                logger.exception("Unexpected error while mounting new workspace")
+
+        mount_nursery.start_soon(target)
+
+    # Instantiate the mountpoint manager with its own nursery
     async with trio.open_service_nursery() as nursery:
         mountpoint_manager = MountpointManager(
             user_fs, event_bus, base_mountpoint_path, config, runner, nursery
         )
+
+        # Exit this context by unmounting all mountpoints
         try:
-            yield mountpoint_manager
 
+            # A nursery dedicated to new workspace events
+            async with trio.open_service_nursery() as mount_nursery:
+
+                # Setup new workspace events
+                events = []
+                if mount_on_workspace_created:
+                    events = [("fs.workspace.created", on_new_workspace)]
+                if mount_on_workspace_shared:
+                    events += [("sharing.updated", on_new_workspace)]
```

Comment:
What about umount on sharing lost when `mount_on_workspace_shared=False` ?

Reply:
I was thinking about it too, I'll investigate.